sdk/publicarchive/downloader.go

```diff
@@ -0,0 +1,215 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package publicarchive
+
+import (
+	"archive/tar"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanche-cli/sdk/network"
+	"github.com/ava-labs/avalanchego/utils/constants"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/cavaliergopher/grab/v3"
+	"go.uber.org/zap"
+
+	sdkConstants "github.com/ava-labs/avalanche-cli/sdk/constants"
+)
+
+const (
+	updateInterval = 500 * time.Millisecond
+	maxFileSize    = 10 * 1024 * 1024 * 1024 // 10GB per file
+	// public archive
+	PChainArchiveFuji = "https://avalanchego-public-database.avax-test.network/testnet/p-chain/avalanchego/data-tar/latest.tar"
+)
+
+type Getter struct {
+	client        *grab.Client
+	request       *grab.Request
+	size          int64
+	bytesComplete int64
+	mutex         *sync.RWMutex
+}
+
+type Downloader struct {
+	getter    Getter
+	logger    logging.Logger
+	currentOp *sync.Mutex
+}
+
+// newGetter returns a new Getter
+func newGetter(endpoint string, target string) (Getter, error) {
+	if request, err := grab.NewRequest(target, endpoint); err != nil {
+		return Getter{}, err
+	} else {
+		return Getter{
+			client:        grab.NewClient(),
+			request:       request,
+			size:          0,
+			bytesComplete: 0,
+			mutex:         &sync.RWMutex{},
+		}, nil
+	}
+}
+
+// NewDownloader returns a new Downloader
+// network: the network to download from ( fuji or mainnet). todo: add mainnet support
+// target: the path to download to
+// logLevel: the log level
+func NewDownloader(
+	network network.Network,
+	logLevel logging.Level,
```

Comment:
can it receive a logger instead of log level

Reply:
that is to have full control on the logs

Reply:
agree